File: Data-Structures-Algorithms/Range-Query_Prefix-Sum/best-time-to-buy-and-sell-stock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/submissions/?envType=study-plan-v2&envId=top-interview-150

class Solution {
    public int maxProfit(int[] prices) {

        int profit = 0;
        int buy = prices[0];

        for (int i = 1; i < prices.length; i++) {
            int sell = prices[i];
            if (sell > buy) {
                profit = Math.max(profit, sell - buy);
            } else {
                buy = sell;
            }
        }

        return profit;
        
    }
}

File: Data-Structures-Algorithms/Range-Query_Prefix-Sum/best-time-to-buy-and-sell-stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

# https://neetcode.io/problems/buy-and-sell-crypto?list=neetcode150

class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        l = 0
        r = 1
        max_profit = 0

        while r < len(prices):
            if prices[l] < prices[r]:
                curr_profit = prices[r] - prices[l]
                max_profit = max(max_profit, curr_profit)
            else:
                l = r
            r += 1
        
        return max_profit
        